src/Distributed/distributed_fft_based_poisson_solver.jl

```diff
@@ -1,115 +1,183 @@
 import PencilFFTs
+using PencilArrays: Permutation
 
 import Oceananigans.Solvers: poisson_eigenvalues, solve!
 import Oceananigans.Architectures: architecture
 
-struct DistributedFFTBasedPoissonSolver{P, F, L, λ, S, C}
-    # Do we need backwards_plan :: B too?
+struct DistributedFFTBasedPoissonSolver{P, F, L, λ, S, I}
     plan :: P
     global_grid :: F
     local_grid :: L
     eigenvalues :: λ
     storage :: S
-    communicator :: C
+    input_permutation :: I
 end
 
 architecture(solver::DistributedFFTBasedPoissonSolver) =
     architecture(solver.global_grid)
 
+infer_transform(grid, d) = infer_transform(topology(grid, d)())
+infer_transform(::Periodic) = PencilFFTs.Transforms.FFT!()
+infer_transform(::Bounded) = PencilFFTs.Transforms.R2R!(FFTW.REDFT10)
+infer_transform(::Flat) = PencilFFTs.Transforms.NoTransform!()
+
+"""
+    DistributedFFTBasedPoissonSolver(global_grid, local_grid)
+
+Return a FFT-based solver for the Poisson equation,
+
+```math
+∇²x = b
+```
+
+for `MultiArch`itectures.
+
+Supported configurations
+========================
+
+We support two "modes":
+
+    1. Two-dimensional decompositions in (x, y) for problems with either
+       `Nz > Rx` or `Nz > Ry` (therefore, three-dimensional).
+
+    2. One-dimensional decompositions in either x or y for problems that are
+       either two-dimensional, or have limited dimensionality in z.
```

Comment:
What is meant by limited dimensionality in `z`? Could we clarify this a bit?

Reply:
It's defined precisely just above. I'll mention that.